mitmproxy/protocol/http.py

```diff
@@ -187,8 +187,7 @@ def __call__(self):
                     flow.request.set_auth(self.config.upstream_auth)
                 self.process_request_hook(flow)
 
-                if not flow.response:
-                    self.establish_server_connection(flow)
+                if not flow.response and self.establish_server_connection(flow):
```

Comment:
If the http2 layer gets terminated (due to a closed connection), `establish_server_connection` would go into a weird state in the h2 protocol - so it is easier just to cancel the whole stack of layers, and begin a fresh connection.